File: Extensification/Collections/Array/Counts.vb

'    Extensification  Copyright (C) 2020-2021  EoflaOE
'
'    This file is part of Extensification
'
'    Extensification is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Extensification is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.Runtime.CompilerServices

Namespace ArrayExts
    ''' <summary>
    ''' Provides the array extensions related to counting
    ''' </summary>
    Public Module Counts

        ''' <summary>
        ''' Gets how many non-empty items are there
        ''' </summary>
        ''' <typeparam name="T">Type</typeparam>
        ''' <param name="TargetArray">Target array</param>
        ''' <returns>Count of non-empty items</returns>
        <Extension>
        Public Function CountFullEntries(Of T)(TargetArray() As T) As Long
            Dim FullEntries As Long
            For i As Long = 0 To TargetArray.LongLength - 1
                If TargetArray(i) IsNot Nothing Then
                    If TypeOf TargetArray(i) Is String Then
                        If Not TargetArray(i).Equals("") Then
                            FullEntries += 1
                        End If
                    Else
                        FullEntries += 1
                    End If
                End If
            Next
            Return FullEntries
        End Function

        ''' <summary>
        ''' Gets how many empty items are there
        ''' </summary>
        ''' <typeparam name="T">Type</typeparam>
        ''' <param name="TargetArray">Target array</param>
        ''' <returns>Count of empty items</returns>
        <Extension>
        Public Function CountEmptyEntries(Of T)(TargetArray() As T) As Long
            Dim EmptyEntries As Long
            For i As Long = 0 To TargetArray.LongLength - 1
                If TargetArray(i) Is Nothing Then
                    EmptyEntries += 1
                ElseIf TypeOf TargetArray(i) Is String And TargetArray(i).Equals("") Then
                    EmptyEntries += 1
                End If
            Next
            Return EmptyEntries
        End Function

    End Module
End Namespace

File: Extensification/Collections/Array/Counts.cs
using Microsoft.VisualBasic.CompilerServices;

namespace Extensification.ArrayExts
{
    /// <summary>
    /// Provides the array extensions related to counting
    /// </summary>
    public static class Counts
    {

        /// <summary>
        /// Gets how many non-empty items are there
        /// </summary>
        /// <typeparam name="T">Type</typeparam>
        /// <param name="TargetArray">Target array</param>
        /// <returns>Count of non-empty items</returns>
        public static long CountFullEntries<T>(this T[] TargetArray)
        {
            var FullEntries = default(long);
            for (long i = 0L, loopTo = TargetArray.LongLength - 1L; i <= loopTo; i++)
            {
                if (TargetArray[(int)i] is not null)
                {
                    if (Conversions.ToBoolean(TargetArray[(int)i] is string))
                    {
                        if (!TargetArray[(int)i].Equals(""))
                        {
                            FullEntries += 1L;
                        }
                    }
                    else
                    {
                        FullEntries += 1L;
                    }
                }
            }
            return FullEntries;
        }

        /// <summary>
        /// Gets how many empty items are there
        /// </summary>
        /// <typeparam name="T">Type</typeparam>
        /// <param name="TargetArray">Target array</param>
        /// <returns>Count of empty items</returns>
        public static long CountEmptyEntries<T>(this T[] TargetArray)
        {
            var EmptyEntries = default(long);
            for (long i = 0L, loopTo = TargetArray.LongLength - 1L; i <= loopTo; i++)
            {
                if (TargetArray[(int)i] is null)
                {
                    EmptyEntries += 1L;
                }
                else if (Conversions.ToBoolean(TargetArray[(int)i] is string) & TargetArray[(int)i].Equals(""))
                {
                    EmptyEntries += 1L;
                }
            }
            return EmptyEntries;
        }

    }
}